server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java

```diff
@@ -105,7 +109,19 @@ void handleHandshake(TransportChannel channel, long requestId, StreamInput strea
             assert ignoreDeserializationErrors : exception;
             throw exception;
         }
-        channel.sendResponse(new HandshakeResponse(this.version));
+        if ((request.version != null) && (request.version.compareTo(this.minAcceptedVersion) < 0)) {
```

Comment:
Why is this necessary? The logic you are pursuing is implemented in `TransportService` handshake. I don't see the need to change the behavior of this handshake which serves a single purpose (negotiate transport protocol version).

Additionally as far as I can tell since this setting is not updatable which depending on the versions set could prevent a cluster form upgrading?

Reply:
The goal is to check the min version on both sides of a handshake request/response, to prevent a cluster from forming. It takes precedence over minCompatibilityVersion, or defaults to minCompatibilityVersion if not set.

TransportService sends a handshake request, and checks the response version against minAcceptedVersion.
https://github.com/elastic/elasticsearch/blob/a194f8d33b4006e01fc45a0d4f52a3af4856230f/server/src/main/java/org/elasticsearch/transport/TransportService.java#L561

TransportHandshaker receive a handshake request, and checks the request version against minAcceptedVersion.
https://github.com/elastic/elasticsearch/blob/a194f8d33b4006e01fc45a0d4f52a3af4856230f/server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java#L112

Reply:
What is the upgrade issue you mentioned? The only example I can think of it if minCompatibilityVersion gets bumped.

Theoretical Future Example:
- **Before upgrade**: currentVersion=`8.6.0`, minCompatibilityVersion=`7.17.0`, minAllowedVersion=`7.17.0`
- **During rolling upgrade**: currentVersion=`9.0.0`, minCompatibilityVersion=`8.6.0`, minAllowedVersion=`7.17.0` (rejected)

I can change the lower limit check of minAllowedVersion from an exception to a log message. Is that the use case you are thinking about, or something else?

Reply:
These are different handshakes. `TransportHandshaker` specifically is the protocol version negotiation handler. The handshake in `TransportService` is designed for validating specific node characteristics.

This is the handler for receiving the handshake request: https://github.com/elastic/elasticsearch/blob/a194f8d33b4006e01fc45a0d4f52a3af4856230f/server/src/main/java/org/elasticsearch/transport/TransportService.java#L273

Reply:
@tbrooks8 I don't understand why this is relevant. The code does not make it clear why there are two different handshake senders, and two different HandshakeRequest classes.

Reply:
Resolved. Thank you for clarifying the different between the two handshakes.